### PR TITLE
fix: channel leak

### PR DESCRIPTION
**Pull request by lentil1016**
# Changes

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [`spec` PR link](https://github.com/katanomi/spec) included
- [ ] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [ ] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [ ] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [ ] Release notes block below has been filled in or deleted (only if no user facing changes)

# Release Notes

### CODE REVIEW

File: parallel/task.go [hunk 1]
```diff
@@ -92,8 +92,6 @@ func P(log *zap.SugaredLogger, name string, tasks ...Task) *ParallelTasks {
 		errLock:     &sync.Mutex{},
 		errs:        []error{},
 		Log:         log,
-
-		doneChan: make(chan struct{}),
 	}
 }
```
Comment (lentil1016) [inline parallel/task.go:96]: premature initialization may cause channel leak

File: parallel/task.go [hunk 2]
```diff
@@ -212,7 +212,7 @@ func (p *ParallelTasks) Do() *ParallelTasks {
 
 	for i, task := range p.tasks {
 		if !p.waitThreshold() {
```
Comment (lentil1016) [inline parallel/task.go:214]: This is where channel initialized. Should ensure they will be release in this func.

File: parallel/task.go [hunk 3]
```diff
@@ -277,11 +282,6 @@ func (p *ParallelTasks) Wait() ([]interface{}, error) {
 		return nil, nil
 	}
 
-	go func() {
-		p.wg.Wait()
-		p.done(nil)
-	}()
-
```
Comment (lentil1016) [inline parallel/task.go:284]: This may cause channel leak
image

File: parallel/task.go [hunk 4]
```diff
@@ -249,6 +249,11 @@ func (p *ParallelTasks) Do() *ParallelTasks {
 
 		}(i, task)
 	}
+	close(p.threshold)
+	go func() {
+		p.wg.Wait()
+		p.done(nil)
+	}()
```
Comment (lentil1016) [inline parallel/task.go:256]: Make sure they will be released once they are initialized.

File: parallel/task.go [hunk 5]
```diff
@@ -249,6 +249,11 @@ func (p *ParallelTasks) Do() *ParallelTasks {
...
 
 	return p
 }
```
Comment (l-qing) [inline parallel/task.go:259]: 这个下面的 close(p.doneChan) 是不是有风险？
因为前面的还未初始化？

Comment (lentil1016) [reply parallel/task.go:259]: 已去除该修改

File: parallel/task.go [hunk 6]
```diff
@@ -250,6 +249,12 @@ func (p *ParallelTasks) Do() *ParallelTasks {
 		}(i, task)
 	}
 
+	go func() {
+		// wait until done, then release channels
```
Comment (chengjingtao) [inline parallel/task.go:253]: 多加一些注释说明下， 实际上是，没调用Wait 的时候 导致了  199 行的 goroutine  一直卡着， goroutine 不释放。

**Review by chengjingtao — DISMISSED** [review]

File: parallel/task.go (hunk 6 above)
Comment (lentil1016) [reply parallel/task.go]: 已补充注释

**Review by chengjingtao — APPROVED** [review]

**Pull request merged**